x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/SharedBytes.java

```diff
@@ -28,136 +35,141 @@
+            // the file up front
+            final ByteBuffer fillBytes = ByteBuffer.allocate(Channels.WRITE_CHUNK_SIZE);
+            this.fileChannel = FileChannel.open(cacheFile, OPEN_OPTIONS);
+            long written = fileChannel.size();
+            fileChannel.position(written);
+            while (written < fileSize) {
+                final int toWrite = Math.toIntExact(Math.min(fileSize - written, Channels.WRITE_CHUNK_SIZE));
+                fillBytes.position(0).limit(toWrite);
+                Channels.writeToChannel(fillBytes, fileChannel);
+                written += toWrite;
+            }
+            if (written > fileChannel.size()) {
+                fileChannel.truncate(fileSize);
+            }
         } else {
             this.fileChannel = null;
         }
+        this.path = cacheFile;
     }
 
-    FileChannel getFileChannel(int sharedBytesPos) {
-        assert fileChannel != null;
-        // return fileChannel;
-        return new FileChannel() {
-            @Override
-            public int read(ByteBuffer dst) throws IOException {
-                throw new UnsupportedOperationException();
-            }
-
-            @Override
-            public long read(ByteBuffer[] dsts, int offset, int length) throws IOException {
-                throw new UnsupportedOperationException();
-            }
+    // TODO: dry up against MLs usage of the same method
+    private static long getUsableSpace(Path path) throws IOException {
+        long freeSpaceInBytes = Environment.getFileStore(path).getUsableSpace();
 
-            @Override
-            public int write(ByteBuffer src) throws IOException {
-                throw new UnsupportedOperationException();
-            }
-
-            @Override
-            @SuppressForbidden(reason = "Use positional writes on purpose")
-            public long write(ByteBuffer[] srcs, int offset, int length) throws IOException {
-                checkOffsets(offset, length);
-                return fileChannel.write(srcs, offset, length);
-            }
-
-            @Override
-            public long position() throws IOException {
-                return fileChannel.position();
-            }
-
-            @Override
-            public FileChannel position(long newPosition) throws IOException {
-                checkOffsets(newPosition, 0);
-                return fileChannel.position(newPosition);
-            }
+        /* See: https://bugs.openjdk.java.net/browse/JDK-8162520 */
+        if (freeSpaceInBytes < 0) {
+            freeSpaceInBytes = Long.MAX_VALUE;
+        }
+        return freeSpaceInBytes;
+    }
 
-            @Override
-            public long size() throws IOException {
-                return fileChannel.size();
-            }
+    @Override
+    protected void closeInternal() {
+        try {
+            IOUtils.close(fileChannel, path == null ? null : () -> Files.deleteIfExists(path));
```

Comment:
I wasn't super sure here, as long as we don't persist the contents of this file in a way that allows for reusing it I figured there was no point in keeping it around across restarts (except for making start-up faster). It also felt safer to delete here. Otherwise configuring a large file and then failing to allocate it (because of concurrent disk space usage changes or so) would cause the disk to be left full even though the node fails to start up.

Reply:
Yeah, we don't need to persist this file across restarts. We should also make sure it's cleaned up when fileSize == 0 after restart

Reply:
++ I pushed 046f9988ade